src/v/iceberg/tests/manifest_serialization_test.cc

```diff
@@ -0,0 +1,253 @@
+
+    iobuf&& release() {
+        buf_.trim_back(buf_.size_bytes() - byte_count_);
+        available_ = 0;
+        byte_count_ = 0;
+        vlog(icelog.trace, "ostream::release({})", buf_.size_bytes());
+        return std::move(buf_);
+    }
+
+    // If there's no available space in the buffer, allocates `chunk_size_`
+    // bytes at the end of the buffer.
+    //
+    // Returns the current position in the buffer, and the available remaining
+    // space.
+    bool next(uint8_t** data, size_t* len) final {
+        vlog(icelog.trace, "ostream::next()");
+        if (available_ == 0) {
+            buf_.append(ss::temporary_buffer<char>{chunk_size_});
+            available_ = chunk_size_;
+        }
+        auto back_frag = buf_.rbegin();
+        *data = reinterpret_cast<uint8_t*>(
+          back_frag->share(chunk_size_ - available_, available_).get_write());
+        *len = available_;
+        byte_count_ += available_;
+        available_ = 0;
+        return true;
+    }
+
+    void backup(size_t len) final {
+        vlog(icelog.trace, "ostream::backup({})", len);
+        available_ += len;
+        byte_count_ -= len;
+    }
+
+    uint64_t byteCount() const final {
+        vlog(icelog.trace, "ostream::bytecount()");
+        return byte_count_;
+    }
+
+    void flush() final {}
+
+private:
+    // Size in bytes with which to allocate new fragments.
+    const size_t chunk_size_;
+
+    iobuf buf_;
+
+    // Bytes remaining in the last fragment in the buffer.
+    size_t available_;
+
+    // Total number of bytes.
+    size_t byte_count_;
+};
+
+// InputStream implementation that takes an iobuf as input.
+// Iterates through the fragments of the given iobuf.
+class avro_iobuf_istream : public avro::InputStream {
+public:
+    explicit avro_iobuf_istream(iobuf buf)
+      : buf_(std::move(buf))
+      , cur_frag_(buf_.begin())
+      , cur_frag_pos_(0)
+      , cur_pos_(0) {}
+
+    // Returns the contiguous chunk of memory and the length of that memory.
+    bool next(const uint8_t** data, size_t* len) final {
+        vlog(icelog.trace, "istream::next()");
+        if (cur_frag_ == buf_.end()) {
+            return false;
+        }
+        auto left_in_frag = cur_frag_->size() - cur_frag_pos_;
+        while (left_in_frag == 0) {
+            ++cur_frag_;
+            cur_frag_pos_ = 0;
+            if (cur_frag_ == buf_.end()) {
+                return false;
+            }
+            left_in_frag = cur_frag_->size();
+        }
+        *data = reinterpret_cast<const uint8_t*>(
+          cur_frag_->share(cur_frag_pos_, left_in_frag).get());
+        *len = left_in_frag;
+        cur_frag_pos_ = cur_frag_->size();
+        cur_pos_ += cur_frag_->size();
+        return true;
+    }
+
+    void backup(size_t len) final {
+        vlog(icelog.trace, "istream::backup({})", len);
+        cur_pos_ -= len;
+        if (cur_frag_ == buf_.end()) {
+            cur_frag_ = std::prev(buf_.end());
```

Comment:
what prevents this from being invoked on buf_ that's empty just part of the InputStream contract?

Reply:
Yeah, pretty much. I don't think the user of this has that well a defined contract, and I wouldn't be surprised if there are some bugs that shake out of this, but the code I'm looking through as a user is here https://github.com/apache/avro/blob/d0d80032671bb8f8c430691e484e9d733a6634fd/lang/c%2B%2B/impl/DataFile.cc